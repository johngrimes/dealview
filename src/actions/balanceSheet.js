// @flow

import { putObject, getObject } from 'db/db'
import type { BalanceSheetOverTime } from 'types/balanceSheet'
import type { AssetMap } from 'types/assets/asset'
import type { LiabilityMap } from 'types/liabilities/liability'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'BalanceSheet'
const theOneAndOnlyKey = 'BalanceSheet'

type CalculateBalanceSheetTask = {
  task: 'CALCULATE_BALANCE_SHEET',
  assets: AssetMap,
  liabilities: LiabilityMap,
  startDate: string,
  endDate: string,
}

type LoadBalanceSheetRequestAction = {
  type: 'LOAD_BALANCE_SHEET_REQUEST'
}
type LoadBalanceSheetSuccessAction = {
  type: 'LOAD_BALANCE_SHEET_SUCCESS',
  balanceSheet: BalanceSheetOverTime
}
type LoadBalanceSheetFailureAction = {
  type: 'LOAD_BALANCE_SHEET_FAILURE',
  error: string|null
}

type UpdateBalanceSheetRequestAction = {
  type: 'UPDATE_BALANCE_SHEET_REQUEST',
  assets: AssetMap,
  liabilities: LiabilityMap,
  startDate: string,
  endDate: string,
}
type UpdateBalanceSheetSuccessAction = {
  type: 'UPDATE_BALANCE_SHEET_SUCCESS',
  balanceSheet: BalanceSheetOverTime,
}
type UpdateBalanceSheetFailureAction = {
  type: 'UPDATE_BALANCE_SHEET_FAILURE',
  error: string|null,
}

type InvalidateBalanceSheetAction = {
  type: 'INVALIDATE_BALANCE_SHEET'
}

export type BalanceSheetAction = LoadBalanceSheetRequestAction
                               | LoadBalanceSheetSuccessAction
                               | LoadBalanceSheetFailureAction
                               | UpdateBalanceSheetRequestAction
                               | UpdateBalanceSheetSuccessAction
                               | UpdateBalanceSheetFailureAction
                               | InvalidateBalanceSheetAction

export const calculateBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap, startDate: string, endDate: string): CalculateBalanceSheetTask => {
  return {
    task: 'CALCULATE_BALANCE_SHEET',
    assets,
    liabilities,
    startDate,
    endDate,
  }
}

export const loadBalanceSheetRequest = (): LoadBalanceSheetRequestAction => {
  return { type: 'LOAD_BALANCE_SHEET_REQUEST' }
}

export const loadBalanceSheetSuccess = (balanceSheet: BalanceSheetOverTime): LoadBalanceSheetSuccessAction => {
  return {
    type: 'LOAD_BALANCE_SHEET_SUCCESS',
    balanceSheet,
  }
}

export const loadBalanceSheetFailure = (error: string|null): LoadBalanceSheetFailureAction => {
  return {
    type: 'LOAD_BALANCE_SHEET_FAILURE',
    error,
  }
}

export const updateBalanceSheetRequest = (assets: AssetMap, liabilities: LiabilityMap, startDate: string, endDate: string): UpdateBalanceSheetRequestAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_REQUEST',
    assets,
    liabilities,
    startDate,
    endDate,
  }
}

export const updateBalanceSheetSuccess = (balanceSheet: BalanceSheetOverTime): UpdateBalanceSheetSuccessAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_SUCCESS',
    balanceSheet,
  }
}

export const updateBalanceSheetFailure = (error: string|null): UpdateBalanceSheetFailureAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_FAILURE',
    error,
  }
}

export const invalidateBalanceSheet = (): InvalidateBalanceSheetAction => {
  return { type: 'INVALIDATE_BALANCE_SHEET' }
}

export const updateBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap,
                                   startDate: string, endDate: string): Thunk => {
  return dispatch => {
    dispatch(updateBalanceSheetRequest(assets, liabilities, startDate, endDate))
    return new Promise(async (resolve, reject) => {
      try {
        const worker = new Worker('/static/js/worker.js', { name: 'Balance Sheet Worker' })
        worker.onmessage = async event => {
          console.log('Balance sheet calc returned:', event.data)
          // flow-ignore
          const saved = await putObject(objectStore, event.data, theOneAndOnlyKey)
          dispatch(updateBalanceSheetSuccess(saved))
          resolve(saved)
        }
        worker.postMessage([ assets, liabilities, startDate, endDate ])
      } catch (error) {
        dispatch(updateBalanceSheetFailure(error))
        reject(error)
      }
    })
  }
}

export const loadBalanceSheet = (): Thunk => {
  return dispatch => {
    dispatch(loadBalanceSheetRequest())
    return new Promise(async (resolve, reject) => {
      try {
        const balanceSheet = await getObject(objectStore, theOneAndOnlyKey)
        dispatch(loadBalanceSheetSuccess(balanceSheet))
        resolve(balanceSheet)
      } catch (error) {
        dispatch(loadBalanceSheetFailure(error))
        reject(error)
      }
    })
  }
}
