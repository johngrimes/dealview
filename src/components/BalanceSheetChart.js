import React from 'react'
import PropTypes from 'prop-types'
import { AreaChart, Area, XAxis, YAxis, Tooltip, CartesianGrid } from 'recharts'
import _ from 'lodash'
import moment from 'moment'

import {
  DateStorageFormat,
  DateDisplayFormat,
  formatDollars,
} from '../data/commonTypes.js'

import './styles/BalanceSheetChart.css'

class BalanceSheetChart extends React.Component {
  static propTypes = {
    balanceSheet: PropTypes.shape({
      status: PropTypes.oneOf([ 'uninitialised', 'loading', 'loaded', 'error' ]),
      objects: PropTypes.objectOf(PropTypes.object),
    }),
  }

  processedData() {
    const data = this.props.balanceSheet
    const dates = _.keys(data)
    const interval = Math.ceil(dates.length / 100) // TODO: Convert 100 into a prop
    const filteredDates = _.filter(dates, (_, i) => i % interval === 0)
    const filteredData = _.pick(data, filteredDates)
    return _.map(filteredData, (v, k) => ({
      ...v,
      date: moment(k, DateStorageFormat).format(DateDisplayFormat),
    }))
  }

  render() {
    const data = this.processedData()
    return (
      <AreaChart data={data} width={900} height={400}>
        <Area type='monotone' dataKey='totalAssets' />
        <XAxis dataKey='date' />
        <YAxis
          dataKey='totalAssets'
          tickFormatter={tick => formatDollars(tick)}
        />
        <CartesianGrid strokeDasharray='3 3' />
        <Tooltip
          formatter={x => {
            console.log(x)
            return x
          }}
        />
      </AreaChart>
    )
  }
}

// class DollarLabel extends React.Component {
//   render() {
//     const { x, y, payload } = this.props
//     return (
//       <text x={x} y={y}>
//         {formatDollars(payload.value)}
//       </text>
//     )
//   }
// }

export default BalanceSheetChart
