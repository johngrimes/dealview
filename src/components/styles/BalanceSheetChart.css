.recharts-cartesian-axis-tick-value {
  font-size: 12px;
}
