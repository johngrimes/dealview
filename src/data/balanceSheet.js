import values from 'lodash.values'
import arrayHead from 'lodash.head'
import arrayTail from 'lodash.tail'
import moment from 'moment'

import { DateStorageFormat } from './commonTypes.js'
import { getValuationAtDate } from './asset.js'
import { getBalanceAtDate } from './loan.js'

export const balanceSheetOverTime = (
  assets,
  liabilities,
  startDate,
  endDate
) => {
  return Object.assign(
    {},
    ...calcBalanceSheet(assets, liabilities, startDate, endDate)
  )
}

function * calcBalanceSheet(assets, loans, date, endDate) {
  let nextDate = date
  while (!nextDate.isAfter(endDate)) {
    const totalAssets = sumAssetValueAtDate(values(assets), nextDate)
    const totalLiabilities = sumLoanBalanceAtDate(values(loans), nextDate)
    yield {
      [nextDate.format(DateStorageFormat)]: {
        totalAssets,
        totalLiabilities,
        equity: totalAssets - totalLiabilities,
      },
    }
    nextDate = moment(nextDate).add(1, 'days')
  }
}

const sumAssetValueAtDate = (assets, date) => {
  if (assets.length === 0) return 0
  const { head, tail } = { head: arrayHead(assets), tail: arrayTail(assets) }
  const headResult = getValuationAtDate(head, date)
  return headResult + sumAssetValueAtDate(tail, date)
}

const sumLoanBalanceAtDate = (loans, date) => {
  if (loans.length === 0) return 0
  const { head, tail } = { head: arrayHead(loans), tail: arrayTail(loans) }
  const headResult = getBalanceAtDate(head, date)
  return headResult + sumLoanBalanceAtDate(tail, date)
}
