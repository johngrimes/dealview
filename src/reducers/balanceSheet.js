export const InitialBalanceSheetState = {
  status: 'uninitialised',
  fresh: true,
  balanceSheet: {},
}

const BalanceSheetReducer = (state = InitialBalanceSheetState, action) => {
  switch (action.type) {
    case 'LOAD_BALANCE_SHEET_REQUEST':
    case 'UPDATE_BALANCE_SHEET_REQUEST':
      return {
        ...state,
        status: 'loading',
      }
    case 'LOAD_BALANCE_SHEET_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        balanceSheet: action.balanceSheet,
      }
    case 'UPDATE_BALANCE_SHEET_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        fresh: true,
        balanceSheet: action.balanceSheet,
      }
    case 'LOAD_BALANCE_SHEET_FAILURE':
    case 'UPDATE_BALANCE_SHEET_FAILURE':
      return {
        ...state,
        status: 'error',
        error: action.error,
      }
    case 'INVALIDATE_BALANCE_SHEET':
      return {
        ...state,
        fresh: false,
      }
    default:
      return state
  }
}

export default BalanceSheetReducer
