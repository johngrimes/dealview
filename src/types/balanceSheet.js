// @flow

import _ from 'lodash'
import moment from 'moment'
import type { Moment } from 'moment'

import { DateFormat } from 'types/commonTypes'
import { getValuationAtDate } from 'types/assets/asset'
import type { Asset, AssetMap } from 'types/assets/asset'
import type { LiabilityMap } from 'types/liabilities/liability'

export type BalanceSheet = {
  totalAssets: number,
  totalLiabilities: number,
  equity: number
}
export type BalanceSheetOverTime = { [date: string]: BalanceSheet }

export const balanceSheetOverTime = (assets: AssetMap, liabilities: LiabilityMap,
                                     startDate: string, endDate: string): BalanceSheetOverTime => {
  const startDay = moment(startDate, DateFormat)
  const endDay = moment(endDate, DateFormat)
  return calcBalanceSheet(assets, liabilities, startDay, endDay)
}

const calcBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap,
                          date: Moment, endDate: Moment): BalanceSheetOverTime => {
  if (date.isAfter(endDate)) return {}
  const totalAssets = sumAssetValueAtDate(_.values(assets), date)
  const totalLiabilities = 0  // sumLiabilityValueAtDate(liabilities, date)
  const nextDate = moment(date).add(1, 'days')
  return { ...calcBalanceSheet(assets, liabilities, nextDate, endDate),
    [date.format(DateFormat)]: {
      totalAssets,
      totalLiabilities,
      equity: (totalAssets - totalLiabilities)
    }
  }
}

const sumAssetValueAtDate = (assets: Asset[], date: string): number => {
  if (assets.length === 0) return 0
  const { head, tail } = { head: _.head(assets), tail: _.tail(assets) }
  return getValuationAtDate(head, date) + sumAssetValueAtDate(tail, date)
}
