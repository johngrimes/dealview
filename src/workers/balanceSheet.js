// @flow

import moment from 'moment'

import { balanceSheetOverTime } from 'types/balanceSheet'
import { DateStorageFormat } from 'types/commonTypes'

onmessage = event => {
  console.log(`Balance sheet calc started:`, event.data)
  // flow-ignore
  postMessage(balanceSheetOverTime(
    event.data[0],
    event.data[1],
    moment(event.data[2], DateStorageFormat),
    moment(event.data[3], DateStorageFormat),
  ))
}
