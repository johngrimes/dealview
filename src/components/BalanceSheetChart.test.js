import React from 'react'
import { shallow } from 'enzyme'

import BalanceSheetChart from './BalanceSheetChart.js'
import { validBalanceSheet1 } from '../data/fixtures/balanceSheet.js'

describe('BalanceSheetChart', () => {
  it('should render', () => {
    const props = {
      balanceSheet: validBalanceSheet1,
    }
    const wrapper = shallow(<BalanceSheetChart {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
