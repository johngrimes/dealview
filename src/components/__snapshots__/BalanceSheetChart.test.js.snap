// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalanceSheetChart should render 1`] = `
<AreaChart
  barCategoryGap="10%"
  barGap={4}
  data={
    Array [
      Object {
        "date": "1 Jan 2017",
        "equity": 155070,
        "totalAssets": 1000070,
        "totalLiabilities": 845000,
      },
      Object {
        "date": "2 Jan 2017",
        "equity": 165080,
        "totalAssets": 1000080,
        "totalLiabilities": 835000,
      },
    ]
  }
  height={400}
  layout="horizontal"
  margin={
    Object {
      "bottom": 5,
      "left": 5,
      "right": 5,
      "top": 5,
    }
  }
  stackOffset="none"
  width={900}
>
  <Area
    activeDot={true}
    animationBegin={0}
    animationDuration={1500}
    animationEasing="ease"
    connectNulls={false}
    dataKey="totalAssets"
    dot={false}
    fill="#3182bd"
    fillOpacity={0.6}
    hide={false}
    isAnimationActive={true}
    legendType="line"
    points={Array []}
    stroke="#3182bd"
    type="monotone"
    xAxisId={0}
    yAxisId={0}
  />
  <XAxis
    allowDataOverflow={false}
    allowDecimals={true}
    dataKey="date"
    domain={
      Array [
        0,
        "auto",
      ]
    }
    height={30}
    hide={false}
    mirror={false}
    orientation="bottom"
    padding={
      Object {
        "left": 0,
        "right": 0,
      }
    }
    reversed={false}
    scale="auto"
    tickCount={5}
    type="category"
    width={0}
    xAxisId={0}
  />
  <YAxis
    allowDataOverflow={false}
    allowDecimals={true}
    dataKey="totalAssets"
    domain={
      Array [
        0,
        "auto",
      ]
    }
    height={0}
    hide={false}
    mirror={false}
    orientation="left"
    padding={
      Object {
        "bottom": 0,
        "top": 0,
      }
    }
    reversed={false}
    scale="auto"
    tickCount={5}
    tickFormatter={[Function]}
    type="number"
    width={60}
    yAxisId={0}
  />
  <CartesianGrid
    fill="none"
    height={0}
    horizontal={true}
    horizontalPoints={Array []}
    stroke="#ccc"
    strokeDasharray="3 3"
    vertical={true}
    verticalPoints={Array []}
    width={0}
    x={0}
    y={0}
  />
  <Tooltip
    active={false}
    animationDuration={400}
    animationEasing="ease"
    coordinate={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    cursor={true}
    cursorStyle={Object {}}
    formatter={[Function]}
    isAnimationActive={true}
    itemSorter={[Function]}
    itemStyle={Object {}}
    labelStyle={Object {}}
    offset={10}
    separator=" : "
    viewBox={
      Object {
        "x1": 0,
        "x2": 0,
        "y1": 0,
        "y2": 0,
      }
    }
    wrapperStyle={Object {}}
  />
</AreaChart>
`;
